sys/include/net/gcoap.h

```diff
@@ -612,16 +612,60 @@ typedef struct {
 typedef ssize_t (*gcoap_link_encoder_t)(const coap_resource_t *resource, char *buf,
                                         size_t maxlen, coap_link_encoder_ctx_t *context);
 
+/**
+ * @name    Return values for resource related operations
+ * @{
+ */
+#define GCOAP_RESOURCE_FOUND        (0)
+#define GCOAP_RESOURCE_MISMATCH     (1)
+#define GCOAP_RESOURCE_WRONG_METHOD (2)
+#define GCOAP_RESOURCE_NO_PATH      (3)
+#define GCOAP_RESOURCE_ERROR        (4)
+/** @} */
+
+/**
+ * @brief   Forward declaration of the gcoap listener state container
+ */
+typedef struct gcoap_listener gcoap_listener_t;
+
+/**
+ * @brief   Handler function for the request matcher strategy
+ *
+ * @param[in]  listener     Listener context
+ * @param[out] resource     Matching resource
+ * @param[in]  pdu          Pointer to the PDU
+ *
+ * @return  GCOAP_RESOURCE_FOUND      on resource match
+ * @return  GCOAP_RESOURCE_MISMATCH   on resource mismatch
+ * @return  GCOAP_RESOURCE_NO_PATH    on no path in @p pdu
```

Comment:
How is no path being present an error?

NO_PATH is (at least for the response) treated as 4.04 Not Found, so this might say

```
 * @return  GCOAP_RESOURCE_NO_PATH    if @p pdu falls within the matcher's patterns, but for the concrete path given, there is no resource.
```

Reply:
https://github.com/RIOT-OS/RIOT/pull/14029/files#diff-ded782b26b569cf2266e68098c3500e2cbd358ad2a3d161617ea6dd6acee76c7L459

At least for the old bevhior, a request without an uri is considered to be an error.

Reply:
Reading up on your comment below, I noticed that a return code of `<= 0` does not mean "no path", but is an indication of insufficient space.